toolchain/check/facet_type.cpp

```diff
@@ -293,43 +293,55 @@ auto ResolveRewriteConstraintsAndCanonicalize(Context& context,
     return;
   }
 
-  for (size_t i :
-       llvm::seq<size_t>(facet_type.rewrite_constraints.size() - 1)) {
+  for (size_t i = 0; i < facet_type.rewrite_constraints.size() - 1; ++i) {
     auto& constraint = facet_type.rewrite_constraints[i];
     if (constraint.lhs_id == SemIR::ErrorInst::InstId ||
         constraint.rhs_id == SemIR::ErrorInst::InstId) {
       continue;
     }
 
-    if (!context.insts().Is<SemIR::ImplWitnessAccess>(constraint.lhs_id)) {
+    auto lhs_access =
+        context.insts().TryGetAs<SemIR::ImplWitnessAccess>(constraint.lhs_id);
+    if (!lhs_access) {
       continue;
     }
-    auto lhs_access =
-        context.insts().GetAs<SemIR::ImplWitnessAccess>(constraint.lhs_id);
-    if (!context.insts().Is<SemIR::LookupImplWitness>(lhs_access.witness_id)) {
+    auto lhs_lookup = context.insts().TryGetAs<SemIR::LookupImplWitness>(
+        lhs_access->witness_id);
+    if (!lhs_lookup) {
       continue;
     }
-    auto lhs_lookup =
-        context.insts().GetAs<SemIR::LookupImplWitness>(lhs_access.witness_id);
-    if (!IsPeriodSelf(context, lhs_lookup.query_self_inst_id)) {
+    if (!IsPeriodSelf(context, lhs_lookup->query_self_inst_id)) {
       continue;
     }
 
-    auto& next = facet_type.rewrite_constraints[i + 1];
-    if (next.lhs_id != SemIR::ErrorInst::InstId &&
-        next.rhs_id != SemIR::ErrorInst::InstId) {
-      if (constraint.lhs_id == next.lhs_id) {
+    for (auto j : llvm::seq(i + 1, facet_type.rewrite_constraints.size())) {
+      auto& next = facet_type.rewrite_constraints[j];
+      if (constraint.lhs_id != next.lhs_id) {
+        break;
+      }
```

Comment:
Had you considered continuing to loop on `i`:

```suggestion
    while (i + 1 < facet_type.rewrite_constraints.size()) {
      auto& next = facet_type.rewrite_constraints[i + 1];
      if (constraint.lhs_id != next.lhs_id) {
        break;
      }
      ++i;
```

Reply:
Done, worded slightly differently. It is more clear, thanks, I hadn't thought of it.